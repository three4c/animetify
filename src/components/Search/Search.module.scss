.Search {
  position: relative;

  & &__input {
    border: none;
    border-bottom: 2px solid #eee;
    font-size: 3.2rem;
    font-weight: 700;
    padding: 8px 56px 8px 0;
    width: 100%;
    height: 56px;

    &::placeholder {
      color: #666;
      font-style: italic;
    }
  }

  &::before {
    background: linear-gradient(
      90deg,
      rgba(255, 255, 255, 0),
      rgba(255, 255, 255, 1)
    );
    content: '';
    position: absolute;
    top: 2px;
    right: 56px;
    width: 56px;
    height: 52px;
  }
  
  & &__button {
    position: absolute;
    top: 0;
    right: 0;
    height: 100%;

    &Item {
      position: relative;
      height: 54px;
      width: 54px;

      &:not(:last-child)::before {
        background-color: #eee;
        content: '';
        position: absolute;
        top: 0;
        bottom: 0;
        margin: auto 0;
        right: 0;
        height: 32px;
        width: 2px;
      }
    }

    .fa-times,
    .fa-search {
      width: 24px;
      height: 24px;
    }
  }
}
