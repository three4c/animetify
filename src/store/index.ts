import { createStore, applyMiddleware } from 'redux';
import logger from 'redux-logger';

export enum SearchActionType {
  SET_SEARCH_TEXT_ACTION = 'SET_SEARCH_TEXT_ACTION',
  SET_SUGGEST_TEXT_ACTION = 'SET_SUGGEST_TEXT_ACTION',
}

export interface GlobalState {
  searchText: string;
  searchYear: string;
  searchSeason: string;
  searchFlag: boolean;
  suggestText: string;
}

export interface SearchAction extends GlobalState {
  type: SearchActionType;
}

export const setSearchText = (searchText: GlobalState['searchText']) => {
  return {
    type: SearchActionType.SET_SEARCH_TEXT_ACTION,
    searchText,
  };
};

export const setSuggestText = (suggestText: GlobalState['suggestText']) => {
  return {
    type: SearchActionType.SET_SUGGEST_TEXT_ACTION,
    suggestText,
  };
};

export const initialState: GlobalState = {
  searchText: '',
  searchYear: '',
  searchSeason: '',
  searchFlag: false,
  suggestText: '',
};

export const searchReducer = (
  state = initialState,
  action: SearchAction
): GlobalState => {
  switch (action.type) {
    case SearchActionType.SET_SEARCH_TEXT_ACTION:
      /** XXXX Sの正規表現 */
      if (action.searchText.match(/[0-9]{4}\s+[亜-熙-秋]{1}/)) {
        /** 空白で分割 */
        const splitText = action.searchText.split(/\s+/);
        return {
          ...state,
          searchText: '',
          searchYear: splitText[0],
          searchSeason: splitText[1],
          suggestText: '',
          searchFlag: true,
        };
      } else {
        return {
          ...state,
          searchText: action.searchText,
          suggestText: '',
          searchYear: '',
          searchSeason: '',
          searchFlag: true,
        };
      }
    case SearchActionType.SET_SUGGEST_TEXT_ACTION:
      return {
        ...state,
        searchText: '',
        suggestText: action.suggestText,
        searchFlag: false,
      };
    default:
      return state;
  }
};

export const store = createStore(searchReducer, applyMiddleware(logger));
