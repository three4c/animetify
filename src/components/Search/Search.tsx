import React, { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setSearchText, setSuggestText, GlobalState } from '../../store';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch, faTimes } from '@fortawesome/free-solid-svg-icons';

import styles from './Search.module.scss';

const Search: React.FC = () => {
  const store = useSelector<GlobalState, GlobalState>((state) => state);
  const dispath = useDispatch();

  const inputRef = useRef<HTMLInputElement>(null);

  const searchHandler = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (store.suggestText) {
      dispath(setSearchText(store.suggestText));
    }
  };

  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispath(setSuggestText(e.target.value));
  };

  const resetHandler = () => {
    dispath(setSearchText(''));
    dispath(setSuggestText(''));

    inputRef.current?.focus();
  };

  return (
    <form className={styles.Search} onSubmit={searchHandler} role="search">
      <label className="util-hidden" htmlFor="search">
        アニメタイトル、アニメソングタイトル、「2011
        春」のように年と季節からアニメを検索するフォーム
      </label>
      <input
        id="search"
        type="text"
        className={`${styles.Search__input} util-focus`}
        placeholder="アニメタイトルまたは年と季節を入力"
        onChange={changeHandler}
        autoComplete="off"
        ref={inputRef}
        value={
          store.searchFlag
            ? store.searchText
              ? store.searchText
              : `${store.searchYear} ${store.searchSeason}`
            : store.suggestText
        }
      />
      <div className={styles.Search__button}>
        {(store.searchText ||
          store.suggestText ||
          (store.searchYear && store.searchSeason)) && (
          <button
            onClick={resetHandler}
            type="button"
            className={`${styles.Search__buttonItem} util-focus`}
            aria-label="入力されているキーワードを削除するボタン"
          >
            <FontAwesomeIcon icon={faTimes} />
          </button>
        )}
        <button
          className={`${styles.Search__buttonItem} util-focus`}
          aria-label="入力されたキーワードを検索するボタン"
        >
          <FontAwesomeIcon icon={faSearch} />
        </button>
      </div>
    </form>
  );
};

export default Search;
