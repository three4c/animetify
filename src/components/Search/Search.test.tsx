import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { GlobalState } from '../../store';

import { shallow } from 'enzyme';

import Search from './Search';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const useSelectorMock = useSelector as jest.Mock;
const useDispatchMcok = useDispatch as jest.Mock;

const mockState: GlobalState = {
  searchText: '電波女と青春男',
  searchYear: '',
  searchSeason: '',
  searchFlag: false,
  suggestText: '',
};

describe('<Search />', () => {
  it('SnapShot - searhFlag:false', () => {
    useSelectorMock.mockImplementation((callback: any) => callback(mockState));
    useDispatchMcok.mockReturnValue(jest.fn());
    expect(shallow(<Search />)).toMatchSnapshot();
  });

  it('SnapShot(searchText) - searhFlag:true', () => {
    mockState.searchFlag = true;
    useSelectorMock.mockImplementation((callback: any) => callback(mockState));
    useDispatchMcok.mockReturnValue(jest.fn());
    expect(shallow(<Search />)).toMatchSnapshot();
  });

  it('SnapShot(yeaer && season) - searhFlag:true', () => {
    mockState.searchFlag = true;
    mockState.searchText = '';
    mockState.searchYear = '2011';
    mockState.searchSeason = '春';
    useSelectorMock.mockImplementation((callback: any) => callback(mockState));
    useDispatchMcok.mockReturnValue(jest.fn());
    expect(shallow(<Search />)).toMatchSnapshot();
  });
});
