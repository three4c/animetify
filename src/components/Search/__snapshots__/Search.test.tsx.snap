// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> SnapShot - searhFlag:false 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
  role="search"
>
  <label
    className="util-hidden"
    htmlFor="search"
  >
    アニメタイトル、アニメソングタイトル、「2011 春」のように年と季節からアニメを検索するフォーム
  </label>
  <input
    autoComplete="off"
    className="Search__input util-focus"
    id="search"
    onChange={[Function]}
    placeholder="アニメタイトルまたは年と季節を入力"
    type="text"
    value=""
  />
  <div
    className="Search__button"
  >
    <button
      aria-label="入力されているキーワードを削除するボタン"
      className="Search__buttonItem util-focus"
      onClick={[Function]}
      type="button"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f00d",
              "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
    <button
      aria-label="入力されたキーワードを検索するボタン"
      className="Search__buttonItem util-focus"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
            ],
            "iconName": "search",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
  </div>
</form>
`;

exports[`<Search /> SnapShot(searchText) - searhFlag:true 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
  role="search"
>
  <label
    className="util-hidden"
    htmlFor="search"
  >
    アニメタイトル、アニメソングタイトル、「2011 春」のように年と季節からアニメを検索するフォーム
  </label>
  <input
    autoComplete="off"
    className="Search__input util-focus"
    id="search"
    onChange={[Function]}
    placeholder="アニメタイトルまたは年と季節を入力"
    type="text"
    value="電波女と青春男"
  />
  <div
    className="Search__button"
  >
    <button
      aria-label="入力されているキーワードを削除するボタン"
      className="Search__buttonItem util-focus"
      onClick={[Function]}
      type="button"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f00d",
              "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
    <button
      aria-label="入力されたキーワードを検索するボタン"
      className="Search__buttonItem util-focus"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
            ],
            "iconName": "search",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
  </div>
</form>
`;

exports[`<Search /> SnapShot(submit) 1`] = `
<Search>
  <form
    className="Search"
    onSubmit={[Function]}
    role="search"
  >
    <label
      className="util-hidden"
      htmlFor="search"
    >
      アニメタイトル、アニメソングタイトル、「2011 春」のように年と季節からアニメを検索するフォーム
    </label>
    <input
      autoComplete="off"
      className="Search__input util-focus"
      id="search"
      onChange={[Function]}
      placeholder="アニメタイトルまたは年と季節を入力"
      type="text"
      value="2011 春"
    />
    <div
      className="Search__button"
    >
      <button
        aria-label="入力されているキーワードを削除するボタン"
        className="Search__buttonItem util-focus"
        onClick={[Function]}
        type="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                352,
                512,
                Array [],
                "f00d",
                "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
              ],
              "iconName": "times",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-times fa-w-11 "
            data-icon="times"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 352 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </button>
      <button
        aria-label="入力されたキーワードを検索するボタン"
        className="Search__buttonItem util-focus"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f002",
                "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
              ],
              "iconName": "search",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-search fa-w-16 "
            data-icon="search"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </button>
    </div>
  </form>
</Search>
`;

exports[`<Search /> SnapShot(yeaer && season) - searhFlag:true 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
  role="search"
>
  <label
    className="util-hidden"
    htmlFor="search"
  >
    アニメタイトル、アニメソングタイトル、「2011 春」のように年と季節からアニメを検索するフォーム
  </label>
  <input
    autoComplete="off"
    className="Search__input util-focus"
    id="search"
    onChange={[Function]}
    placeholder="アニメタイトルまたは年と季節を入力"
    type="text"
    value="2011 春"
  />
  <div
    className="Search__button"
  >
    <button
      aria-label="入力されているキーワードを削除するボタン"
      className="Search__buttonItem util-focus"
      onClick={[Function]}
      type="button"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f00d",
              "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
    <button
      aria-label="入力されたキーワードを検索するボタン"
      className="Search__buttonItem util-focus"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
            ],
            "iconName": "search",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
  </div>
</form>
`;
